Blum 3-approximate shortest common superstring algorithm

MaikuZ opened the pull request:
Implementacja algorytmu 3-przybliżonego dla problemu najkrótszego nadsłowa wraz z opisem algorytmu i głównym dowodem otrzymanej jakości rozwiązania.

krzysztof-turowski commented on text/MichałZwonek/opis.tex:
```diff
@@ -0,0 +1,173 @@
+\begin{theorem}
+\textbf{TGREEDY} dla zbioru $S = \{s_1, ..., s_m \}$ zwraca nadsłowo $s$ nad $S$ o długości co najwyżej $3\cdot OPT(S)$.
+\end{theorem}
+\begin{proof}
```
For your information: I think that a bit simpler proof you may found in Vazirani (Approximation Algorithms) or in https://www.mimuw.edu.pl/~mucha/teaching/aa2008/ss.pdf - but this one is altogether sufficient.

MaikuZ replied:
Thank you for the information, I'll take a look at the proof you provided. Thanks!

krzysztof-turowski commented on test/test_tgreedy.py:
```diff
@@ -0,0 +1,62 @@
+import unittest
+
+from superstring import tgreedy
+from common import prefix
```
Unused import

MaikuZ commented on superstring/tgreedy.py:
```diff
@@ -0,0 +1,55 @@
+  while S:
+    best_ov = -1
+    chosen_pair = ('','')
+    for w_1 in S:
+      for w_2 in S:
```
I believe that is not what we actually need. Combinations(['aab', 'baa'], 2) would only yield [('aab', 'baa')], whereas we need [('aab', 'baa'), ('baa', 'aab')], only the latter pair would result in an overlap of 2. I think what we need in this case is actually product(S, 2) and in the other case permutations(S, 2). What do you think?

krzysztof-turowski replied:
Yes, you're right, thank you for correction.

MaikuZ commented on text/MichałZwonek/opis.tex:
```diff
@@ -0,0 +1,173 @@
+
+Ponadto Blum et al. zaprezentowali alternatywny algorytm, nazywany \textbf{TGREEDY}, dla którego udowodnili, że jest on 3-aproksymacyjny. Sam problem znalezienia najkrótszego wspólnego nadsłowa jest NP trudny.
+
+W pierwszej sekcji przedstawimy podstawowe definicje i notacje, a w kolejnych algorytm \textbf{MGREEDY}, a w następynej sekcji algorytm 
```
There is a description of greedy in the following part:

"Oczywistym podejściem jest następujący algorytm zachłanny. Po kolei łącz słowa mające największą część wspólną nachodzącą na siebie określaną jako _overlap_. Rób tak, aż do uzyskania 
jednego słowa. Postawiona jest hipoteza, że takie podejście ma współczynnik aproksymacji równy 2, ale przed opublikowaniem pracy przez Blum et al. nie było wiadome, czy taki algorytm jest w ogóle liniową aproksymacją. Okazuje się, że ten algorytm (określać będziemy go poprzez **GREEDY**) ma współczynnik aproksymacji co najwyżej 4."

However, I agree that the mention is easy to miss, so maybe I could rephrase the part you highlighted to something like this:

"W pierwszej sekcji przedstawimy podstawowe definicje i notacje, a w kolejnych algorytm **MGREEDY** będący pewną modyfikacją algorytmu **GREEDY**, taką dla którego analiza aproksymacji jest prostsza. W następnej sekcji 
przedstawimy algorytm **TGREEDY** wykorzystujący algorytm **MGREEDY** by uzyskać pożądany współczynnik aproksymacji. W ostatniej sekcji postaramy się udzielić odpowiedzi na pytanie, który z tych dwóch algorytmów jest lepszy?"

And maybe add something like this to the previous extract:

"[...]Okazuje się, że ten algorytm (określać będziemy go poprzez **GREEDY**) ma współczynnik aproksymacji co najwyżej 4, dowodu tego faktu nie będziemy przybliżać. Sam algorytm jest kluczowy, bo jest punktem startowym dla kolejnych dwóch algorytmów na których w tej pracy się skupimy."

What do you think, does that sound better?

Truth be told, the hardest part in writing the description of the original paper was actually compressing it to 3-4 pages, so that it was still readable, contained the main proof and did not exceed 4 pages. I might have misinterpreted this guideline as a hard upper bound, where it actually was more of an lower bound or a soft upper bound. That is my mistake and I think it might have resulted in poorer reader experience, sorry for that.

krzysztof-turowski replied:
We may leave it like that. I know that's an advice as an afterthought but in such cases do not hesitate to ask.

I'll try to clean this up when I'll be doing the pass to make the notations of all your notes and codes uniform (as that was taken as a soft bound instead, a bit over my expectations). I guess I'll make these three separate sections: greedy (known to be 4-opt), mgreedy (4-opt, but different), tgreedy (3-opt).

MaikuZ replied:
Okay, I'll fix the other issues you've already pointed out and then I think I should have the updated PR ready before midnight.

MaikuZ commented on text/MichałZwonek/opis.tex:
```diff
@@ -0,0 +1,173 @@
+\section*{Uwagi końcowe}
+
+Blum et al. pokazali dodatkowo, że \textbf{GREEDY} osiąga współczynnik aproksymacji równy 4, ale hipoteza czy ten współczynnik można obniżyć do 2 jest wciąż otwarta. 
+Możnaby się pokusić wysunąć hipotezę, że \textbf{TGREEDY} jest lepszy niż \textbf{GREEDY}. Tak nie jest. Są przykłady zbioru $S$ dla których zarówno jeden jak i drugi wypada lepiej. 
```
Thank you! I figured it would be an interesting addition.

MaikuZ commented on text/MichałZwonek/opis.tex:
```diff
@@ -0,0 +1,173 @@
+$t_1,t_2$ w słowo $t = \langle t_1,t_2\rangle$, czyli, że $t$ ma $first(s)$ jako podsłowo. Wtedy wiemy, że 
+$t = ufirst(s)v$ dla jakichś $u,v$. Skoro $first(s)$ nie może być podsłowem 
+$t_1$ ani $t_2$ (założenie o zbiorze $S$), to $|first(s)| > ov(t_1,t_2)$ oraz 
+$|u| < d(t_1,t_2)$. Czyli, $ov(t_1, s) > ov(t_1, t_2)$, co daje nam sprzeczność z faktem, że $ov(t_1,t_2)$ jest największym.
```
Frankly, I don't really see why this is needed. I think I can rephrase it in a neater way, tell me what you think.

Wtedy wiemy, że dla t_1 = ab, t_2 = bc (takiego, że b jest maksymalne), t = ufirst(s)v  oraz t = abc. Skoro first(s) nie może być podsłowem t_1 ani t_2 (założenie o zbiorze S), to wiemy, że first(s) musi zawierać b jako podsłowo i nachodzić przynajmniej jedną literą na a i c. Z tego wynika, że |u| < |a| = d(t_1,t_2), |v| < |c| oraz |first(s)| > ov(t_1, t_2) = |b|. Teraz możemy zauważyć, że ov(t_1, s) >= ov(t_1, first(s)) >= |b| + 1 > ov(t_1, t_2) = |b|, co stoi w sprzeczności z faktem, że ov(t_1, t_2) miał być największą wartość.

I think now it's a bit more graphical this way. How do you think, does it look better like this?

krzysztof-turowski replied:
OK, that's much cleaner now.

MaikuZ commented:
I have fixed the issues you highlighted and updated the PR. Additionally to the changes you proposed I have also added a test for empty sets of words.

MaikuZ commented on superstring/tgreedy.py:
```diff
@@ -50,6 +41,11 @@ def _mgreedy_first_phase(words):
 
+# achieves 4-approximation
+def mgreedy(words):
+  return next(iter(words), [''])[0] + \
```
I have decided to make it a bit more complicated than the obvious solutios, so that it works even for words being empty. If I just took a wildcard from the first string it would then crash for words being empty. Also, I wanted to use the dummy character from a string inputted by the user as we do not have any prior knowledge about what kind of characters can be used as the dummy character.

krzysztof-turowski replied:
That's fine, as this is an educational library, not an industry one, we are allowed to assume certain conventions consistent with the lecture notes - but from a broader perspective I understand your cautiousness about the corner cases.
